### DISCUSSION TITLE
[GIT101] 關於 git 兩個小問題

### DISCUSSION THREAD

**Discussion by raye0621** [Lidemy/forum#57] [其他問題] [answered]
1. 老師上課有提到如果要到一個 branch 的最新狀態的話就輸入 ` git checkout <branchName>` 那這時的狀態跟 ` git checkout <最新的commitName>` 有甚麼差別呢？

舉例如下：
```
master
    |
    |--branchA-commitA-commitB
```  
以上如果 ` git checkout branchA ` 和 ` git checkout commitB ` Git bash 上顯示檔案都會是一樣的修改狀態，那他們主要差別是什麼呢？
看了自我檢討後有認知到前者應該是切換到 branch 的最新 (最後更改) 的檔案版本，
後者則是切換到該 commit 上，可能是只是要修改或是查看而已，不知道有沒有理解錯誤。
想請問這兩個的差別以及通常是什麼狀況會使用到，謝謝！

2. 如果我再回去 commitA 修改並且 commit 的話是不是會再建立一條由 commit 出發的 branch

舉例如下：
```
master(Afile)
    |
    |--branchA-commitA-commitB
       (Afile) (+Bfile) (+Cfile)
       
    //ps.現在 commitA 裡面是 Afile + Bfile ；commitB 裡面是 Afile + Bfile + Cfile
       
↓↓  checkout 回 commitA 加上 Dfile 再 commit     ↓↓

master(Afile)
    |
    |--branchA-commitA-commitB
       (Afile) (+Bfile) (+Cfile)
                  |
               (+Dfile 再 commit)

↓↓  會變成這樣：commit 完會自動多一條 branch 出來    ↓↓

master(Afile)
    |
    |--branchA-commitA-commitB
    |  (Afile) (+Bfile) (+Cfile)
    |            |
    |  -- branch(commitA)   -    commitD
    |         (Afile + Bfile)          (Afile + Bfile + Dfile)
``` 

    
請問這樣理解是正確的嗎？它會直接自動幫你建立一個 branch 這樣子。

以上兩個問題感謝 ！

**Comment by qoosuperman** [comment]
小弟嘗試回答看看

1. 其實 branch name 可以把它想像成一個貼紙，只是給這個 commit 一個代號，所以 checkout 到這個 commit 點跟 checkout 到這個 branch 的意思是一樣的，但是 branch 本身會跟著 commit 移動，如果你現在在 master branch 上面， commit 點是 a12345，你再做一些改變再 add / commit 一次，會發現你雖然還是在 master 上面，但 commit 點已經前進了
2. 同上面的道理，如果你在一個沒有 branch 的 commit 上面做 commit ，那你就只會跑到下一個 commit 上面，因為並沒有一個 branch name 跟著你

**Reply by raye0621** [reply]
了解！感謝回答，
貼紙的比喻很有畫面

**Comment by JAS0NHUANG** [comment]
我也試著回答看看，先上個圖：
（圖片裡每個寫著 C1、C2……的圓圈代表 commit，大箭頭/指標代表 branch，HEAD 就是 HEAD 檔，下面的一行字就是造成當前狀況所下的指令。
![Git HEAD and Branch](https://i.imgur.com/W6V3Fsh.png)

**Reply by raye0621** [reply]
這圖太屌了，完全可以理解。
謝謝分享。

**Comment by JAS0NHUANG** [comment, marked as answer]
先講答案：
  1. `git checkout <branchName>` HEAD 指向的是 branch，`git checkout <commitHash>` HEAD 會直接指向 commit，所以兩個是不一樣的。
  2. 不會建立 branch，但是 HEAD 會跟著 commit 一直接下去。

這裡的重點就是那個 HEAD（`.git/HEAD` 這個檔案），可以把 HEAD 想像成一個小白板，是真的很像，基本上可以無縫類比，大家自已 `cat .git/HEAD` 看看就知道為什麼了，上面平常會寫著目前所在分支的位置（指向 `.git/refs/heads/` 裡面的其中一個分支檔），要改變上面的內容只要 `git checkout branch1` 就可以把白板上的內容變成 branch1 的位置，`git checkout branch2` 就可以把內容變成 branch2 的位置，然後 Git 就會去問 branch1 或 branch2 現在你們最新的 commit 是什麼，而這些分支檔案裡面也很簡單，就是存一個 commit hash 值；如果今天我直接 `git checkout <commitHash>` 的話，小白板就會直接被寫上 commit hash，然後 Git 就會跳過問 branch 的步驟，直接去找到那個 commit，進入「斷頭」狀態（detached head）。

所以第一個問題的答案是兩個 checkout 的方法有很大的差別，`git checkout <branchName>` 會讓 HEAD 指向一個 branch 檔然後「再」指向某個 commit （像附圖第二個圖示裡 `git checkout B` 的狀況），而 `git checkout <commitHash>` 則會讓 HEAD 直接指向最新的 commit（如附圖上第四個圖示裡面 `git checkout C4` 的狀況，而且就算你 checkout 的 commit hash 跟所在 branch 的最新 commit 一樣也同樣會斷頭）。

接著就是第二個問題啦！在「斷頭」的狀態之下我們還是可以在那個斷頭的 HEAD 上 `add`、`commit`，可是 Git 並不會自動建立一個新 branch，而是一直去更新 HEAD 的指向（平常 HEAD 指向某個 branch 時，每次 commit 是不會造成它的變動的），然後因為這個 HEAD 沒有指向任何一個 branch，所以如果 checkout 到另一個地方（把 HEAD 搬走）的話就比較難找到這些新的 commit 了。

如果想要在上面建立一個新的 branch 該怎麼做呢？就只要下 `git branch <newBranchName> <commitHash>` 就可以了，以附圖為例，只要下 `git branch new C7` 就可以建立另一個名字叫 new 的分支（指標）指向 C7 這個 commit 了。

那到底「斷頭」有什麼用處呢？我覺得可以把它想成：我想要在某個 commit 上做些實驗，成功了再想辦法留下來，不成功就 checkout 回原來的 branch 上不管它了。然後沒人管的「斷頭」commit 們會發生什麼事呢？如果一段時間之後沒有人來把它們撿回去（給他一個 branch 或 cherry pick 之類的），那它就會被 Git 的資源回收機制給回收掉啦！

好啦！報告完畢，Git 的水有點深，但是真的不必這麼追究它的細節，能夠正確完成交作業的流程應該就可以應付大部分的情況了。

註：附圖中用 C1、C2、C3……只是為了方便閱讀，他們原來應該就是那一長串沒人記的起來的 commit hash 值。

**Reply by raye0621** [reply]
了解，原來不會自動生成 branch。
因為一開始不知道兩者的差異，在嘗試玩 git 時都很隨意的 checkout commmit 或是 checkout branch，結果發現  branch 的狀態好像不太一樣才覺得怪怪的，而且還出現什麼 detached HEAD。

雖然後來照著一般流程是不會有問題但還是很好奇， JAS0N 大大居然還有提供 detached HEAD 狀態建立新分支的方法，感謝你～

非常感謝 JAS0N 大大的解惑，清晰易懂！
